src/py2spack/loading.py

```diff
@@ -0,0 +1,257 @@
+"""Utils for downloading packages from PyPI and opening the pyproject.tomls."""
+
+import io
+import tarfile
+from typing import Any, Dict, List, Optional
+
+import requests
+import tomli
+from packaging import version as vn
+
+KNOWN_ARCHIVE_FORMATS = [
+    ".tar",
+    ".tar.gz",
+    ".tar.bz2",
+    ".gz",
+    ".xz",
+    ".bz2",
+]
+
+
+class APIError(Exception):
+    """Error during version lookup or querying of the PyPI JSON API."""
+
+    def __init__(
+        self,
+        msg: str,
+    ):
+        """Initialize error."""
+        super().__init__(msg)
+
+
+def _acceptable_version(version: str) -> Optional[vn.Version]:
+    """Check whether version string can be parsed and has correct format."""
+    try:
+        v = vn.parse(version)
+        # do not support post releases of prereleases etc.
+        if v.pre and (v.post or v.dev or v.local):
+            return None
+        return v
+    except vn.InvalidVersion:
+        return None
+
+
+class PyPILookup:
+    """Performs PyPI API calls and caches package versions.
+
+    The full API response (including available files and metadata) is not cached
+    due to its larger size and sparser usage.
+    """
+
+    def __init__(self) -> None:
+        """Initialize empty PyPILookup."""
+        # cache for package name and corresponding version list
+        self.version_cache: Dict[str, List[vn.Version]] = {}
+
+    def _get(self, name: str) -> dict | APIError:
+        """Load info for the available distribution files from PyPI.
+
+        Available versions are cached.
+        """
+        r = requests.get(
+            f"https://pypi.org/simple/{name}/",
+            headers={"Accept": "application/vnd.pypi.simple.v1+json"},
+            timeout=10,
+        )
+        if r.status_code != 200:
+            if r.status_code == 404:
+                msg = f"Package {name} not found on PyPI (status code 404)"
+                return APIError(msg)
+
+            msg = (
+                f"Error when querying JSON API (status code {r.status_code})."
+                f" Response: {r.text}"
+            )
+            return APIError(msg)
+
+        data: dict = r.json()
+        versions = data["versions"]
+
+        # parse and sort versions
+        parsed_versions = sorted(
+            {vv for v in versions if (vv := _acceptable_version(v))}
+        )
+
+        self.version_cache[name] = parsed_versions
+
+        return data
+
+    def get_versions(self, name: str) -> List[vn.Version]:
+        """Get acceptable versions for package (from cache if possible)."""
+        if name not in self.version_cache:
+            data = self._get(name)
+
+            if isinstance(data, APIError):
+                self.version_cache[name] = []
+
+        return self.version_cache[name]
```

Comment:
A similar caching functionality is already provided by `functools.lru_cache()` or `functools.cache()`. Also, the parsing of the versions doesn't belong to the utility method contacting the server but to the actual method dealing with versions.

```suggestion
    def _get(self, name: str) -> dict | APIError:
        """Load info for the available distribution files from PyPI.

        Available versions are cached.
        """
        r = requests.get(
            f"https://pypi.org/simple/{name}/",
            headers={"Accept": "application/vnd.pypi.simple.v1+json"},
            timeout=10,
        )
        if r.status_code != 200:
            if r.status_code == 404:
                msg = f"Package {name} not found on PyPI (status code 404)"
                return APIError(msg)

            msg = (
                f"Error when querying JSON API (status code {r.status_code})."
                f" Response: {r.text}"
            )
            return APIError(msg)

        data: dict = r.json()
        return data

    @functools.cache
    def get_versions(self, name: str) -> List[vn.Version]:
        """Get acceptable versions for package (from cache if possible)."""
        data = self._get(name) 
        if isinstance(data, APIError):
            return []   ## <----- I don't think it's right to return an empty list instead of an error here...
        versions = data["versions"]

        # parse and sort versions
        parsed_versions = sorted(
            {vv for v in versions if (vv := _acceptable_version(v))}
        )
        return parsed_versions
```